src/aiortc/codecs/vpx.py

```diff
@@ -436,6 +436,275 @@ def __update_config(self) -> None:
+            self.image.stride[p] = frame.planes[p].line_size
+
+        # encode frame
+        flags = 0
+        if force_keyframe:
+            flags |= lib.VPX_EFLAG_FORCE_KF
+        _vpx_assert(
+            lib.vpx_codec_encode(
+                self.codec,
+                self.image,
+                frame.pts,
+                self.timestamp_increment,
+                flags,
+                lib.VPX_DL_REALTIME,
+            )
+        )
+
+        it = ffi.new("vpx_codec_iter_t *")
+        length = 0
+        while True:
+            pkt = lib.vpx_codec_get_cx_data(self.codec, it)
+            if not pkt:
+                break
+            elif pkt.kind == lib.VPX_CODEC_CX_FRAME_PKT:
+                # resize buffer if needed
+                if length + pkt.data.frame.sz > len(self.buffer):
+                    new_buffer = bytearray(length + pkt.data.frame.sz)
+                    new_buffer[0:length] = self.buffer[0:length]
+                    self.buffer = new_buffer
+
+                # append new data
+                self.buffer[length : length + pkt.data.frame.sz] = ffi.buffer(
+                    pkt.data.frame.buf, pkt.data.frame.sz
+                )
+                length += pkt.data.frame.sz
+
+        # packetize
+        payloads = []
+        descr = VpxPayloadDescriptor(
+            partition_start=1, partition_id=0, picture_id=self.picture_id
+        )
+        pos = 0
+        while pos < length:
+            descr_bytes = bytes(descr)
+            size = min(length - pos, PACKET_MAX - len(descr_bytes))
+            payloads.append(descr_bytes + self.buffer[pos : pos + size])
+            descr.partition_start = 0
+            pos += size
+        self.picture_id = (self.picture_id + 1) % (1 << 15)
+
+        timestamp = convert_timebase(frame.pts, frame.time_base, VIDEO_TIME_BASE)
+        logger.warning("encoding frame " + str(timestamp) + " type:" +  str(frame.pict_type) + " force keyframe:" + str(force_keyframe))
+        return payloads, timestamp
+
+    @property
+    def target_bitrate(self) -> int:
+        """
+        Target bitrate in bits per second.
+        """
+        return self.__target_bitrate
+
+    @target_bitrate.setter
+    def target_bitrate(self, bitrate: int) -> None:
+        if self.enable_gcc:
+            vpx_bitrate = self.get_vpx_bitrate(bitrate)
+            if vpx_bitrate is not None:
+                bitrate = vpx_bitrate
+            else:
+                bitrate = int((bitrate * 1000 / 173 + 184971))
+
+            bitrate = max(self.vpx_min_bitrate, min(bitrate, self.vpx_max_bitrate))
+            if bitrate != self.__target_bitrate:
+                self.__target_bitrate = bitrate
+                self.__update_config_needed = True
+
+    def __update_config(self) -> None:
+        self.cfg.rc_target_bitrate = self.__target_bitrate // 1000
+        self.__update_config_needed = False
+
 
 def vp8_depayload(payload: bytes) -> bytes:
```

Comment:
Is the payload descriptor identical between vp8/vp9?

Reply:
I thought the payload descriptor would be the same, but now that I'm looking at the RFCs ([7741](https://www.rfc-editor.org/rfc/rfc7741.html#section-4.2) for VP8 and [this draft](https://datatracker.ietf.org/doc/html/draft-ietf-payload-vp9#section-4.2) for VP9), it looks like they are different. Let me look into this a bit more, it shouldn't be hard to make the adjustments.

Reply:
Okay, this works if both sides (the sender and the receiver) use the same payload format. This means that this implementation is NOT interoperable with a standard WebRTC implementation that uses the official payload format for VP9. This should be completely fine for training purposes.

Reply:
My test code:

```python3
import os
import sys
import numpy as np
import matplotlib.pyplot as plt
import aiortc
import av
import importlib

from aiortc.codecs import vpx
from aiortc.jitterbuffer import JitterFrame

container = av.open(file='video.mp4', format=None, mode='r')
stream = container.streams.video[0]

encoder = vpx.Vp9Encoder()
decoder = vpx.Vp9Decoder()

for av_frame in container.decode(stream):
    payloads, ts = encoder.encode(av_frame, quantizer=52, enable_gcc=False)
    payload = b"".join([vpx.vp8_depayload(p) for p in payloads])
    decoded_frame_av = decoder.decode(JitterFrame(data=payload, timestamp=ts))
    f = decoded_frame_av[0].to_rgb().to_ndarray()
    plt.imshow(f)
    print(len(payload))
```

Reply:
Yeah, this may not be an immediate issue, but something worth looking into for end-to-end evaluation.

Reply:
Thank you for clarifying this, and sending me the test code.